### PR TITLE
eth/traces: add state limit

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR introduces a new mechanism in chain tracer for preventing creating too many trace states.

The workflow of chain tracer can be divided into several parts:
- state creator generates trace state in a thread
- state tracer retrieves the trace state and applies the tracing on top in another thread
- state collector gathers all result from state tracer and stream to users

It's basically a producer-consumer model here, while if we imagine that the state producer generates states too fast, then it will lead to accumulate lots of unused states in memory. Even worse, in path-based state scheme it will only keep the latest 128 states in memory, and the newly generated state will invalidate the oldest one by marking it as stale.

The solution for fixing it is to limit the speed of state generation. If there are over 128 states un-consumed in memory, then the creation will be paused until the states are be consumed properly.

### CODE REVIEW

File: eth/tracers/tracker.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		head := t.head
+		t.lock.RUnlock()
+
+		if number < t.head {
+			return fmt.Errorf("invalid state number %d head %d", number, t.head)
+		}
+		if int(number-head) < t.limit {
+			return nil
+		}
+		time.Sleep(time.Millisecond * 100)
```
Comment (holiman) [inline eth/tracers/tracker.go:91]: This is essentially a spinlock. A different way to do it would be to use a `sync.Cond` and do `Sleep`, and do a `Signal` or `Broadcast` on the side that triggers the wakeup.

File: eth/tracers/tracker.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	"time"
+)
+
+// stateTracker is an auxiliary tool used to cache the release functions of all
+// used trace states, and to determine whether the creation of trace state needs
+// to be paused in case there are too many states waiting for tracing.
+type stateTracker struct {
+	limit    int                // Maximum number of states allowed waiting for tracing
+	head     uint64             // The number of the first trace state which isn't used up
+	used     []bool             // List of flags indicating whether the trace state has been used up
```
Comment (holiman) [inline eth/tracers/tracker.go:31]: I'd appreciate some more docs about how `head` and `used` are related.

**Comment by rjl493456442** [comment]
@holiman updated, please take another look

File: eth/tracers/tracker.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	// the distance between the specified state and the oldest state
+	// which is still using for trace.
+	t.used[int(number-t.oldest)] = true
+
+	// If the oldest state is used up, update the oldest marker by moving
+	// it to the next state which is not used up.
+	if number == t.oldest {
+		var count int
+		for i := 0; i < len(t.used); i++ {
+			if !t.used[i] {
```
Comment (holiman) [inline eth/tracers/tracker.go:65]:
```suggestion
		for _, used := range t.used {
			if !used {
```

File: eth/tracers/tracker.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+				break
+			}
+			count += 1
+		}
+		t.oldest += uint64(count)
+		copy(t.used, t.used[count:])
+
+		// The tail of t.used is left unchanged. It will be overwritten
+		// by the following states anyway.
```
Comment (holiman) [inline eth/tracers/tracker.go:74]: This is not obvious to me. Perhaps you can clarify what you mean by "following states" ?

File: eth/tracers/tracker.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+// wait blocks until the accumulated trace states are less than the limit.
+func (t *stateTracker) wait(number uint64) error {
+	t.lock.Lock()
+	defer t.lock.Unlock()
+
+	for {
+		if number < t.oldest {
+			return fmt.Errorf("invalid state number %d head %d", number, t.oldest)
+		}
+		if int(number-t.oldest) < t.limit {
```
Comment (holiman) [inline eth/tracers/tracker.go:102]:
```suggestion
		if number < t.oldest+uint64(t.limit) {
			// number is now within limit, wait over
```
IMO clearer way to express what we want to check. Also keeps it in uint64-space.

**Review by holiman — COMMENTED** [review]
LGTM, I've reviewed it pretty thoroughly, but haven't actually _tested_ it. Will do that too

**Review by holiman — APPROVED** [review]

File: eth/tracers/tracker.go (hunk 4 above)
Comment (rjl493456442) [reply eth/tracers/tracker.go:74]: For example, the size of used array is 5, and the current state os this array is [true, true, false, true, false]
after the loop, we find that the first two flags are marked as true, and we left-shift the array by 2, it's [false, true, false, true, false]

You can see that the last two elements are left unchanged. And actually they are useless dirty data. I didn't explicitly clean them up since although they are wrong dirty data, but they don't belong to any state. 

Whenever we set the flag for the next state, we will overwrite the value of these elements.

**Comment by holiman** [comment]
I guess this is the part I don't find obvious:> Whenever we set the flag for the next state, we will overwrite the value of these elements.

**Comment by rjl493456442** [comment]
@holiman in my example, the last two elements in `used` array are completely dangling(doesn't point to any state), so next time we call `t.used[int(number-t.oldest)] = true` by calling `releaseState`, these elements will be re-assigned to the new states and overwrite the original value in the array. So we don't need to explicitly clean them.

File: eth/tracers/tracker.go (hunk 4 above)
Comment (holiman) [reply eth/tracers/tracker.go:74]: How about another example... 

For example, the size of used array is 5, and the current state os this array is [true, true, false, true, true]
after the loop, we find that the first two flags are marked as true, and we left-shift the array by 2, it's [false, true, true, true, true]

You can see that the last two elements are left unchanged (`true`). And actually they are useless dirty data. 


Now, what I'm curious about is what happens when  the first `false` is turned into `true`. Suddenly all states are flushed as done?

Comment (holiman) [reply eth/tracers/tracker.go:74]: ... To clarify: from that point onward, nothing will ever become `false` inside the `t.used` slice again. As far as I can tell. And that doesn't seem right.

Comment (holiman) [reply eth/tracers/tracker.go:74]: Hm. I think the dissonance has been that I've seen this as data type which 'moves' along with the state generation. But, in fact, it's meant to be used once? 
So you init it with the full range that you expect to ever need? And then you never extend it beyond that? 

Because if so, wouldn't be better to.. 

- Instead of moving subslices to the left, in `t.used`, 
- keep one single index of the first non-true element is

Comment (holiman) [reply eth/tracers/tracker.go:74]: Nope, strike that. I still believe there's something wrong here. At least I can't quite make sense of how this works, if it works :)

File: eth/tracers/tracker_test.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		{
+			limit:   3,
+			calls:   []uint64{1, 2, 0},
+			expHead: 3,
+		},
+		{
+			limit:   3,
+			calls:   []uint64{1, 2, 0, 5, 4, 3},
+			expHead: 6,
+		},
```
Comment (holiman) [inline eth/tracers/tracker_test.go:49]:
```
		{
			limit:   5,
			calls:   []uint64{4, 3, 2, 1, 0},
			expHead: 5,
		},
		{
			limit:   5,
			calls:   []uint64{0, 1, 2, 3, 4},
			expHead: 5,
		},

```
--> 
```
[true true true true true]
[false false false false false]
```

Comment (holiman) [reply eth/tracers/tracker_test.go:53]: With the most recent change, this now becomes 
```
    tracker_test.go:77: used: [false false false false false]
    tracker_test.go:77: used: [false false false false false]
```
:+1:

**Comment by holiman** [comment]
One remaining thought. So we have 
1. one producer routine to feed all the blocks into the tracers
2. `threads` consumer routines, doing tracing. 

The producer may (now) get stuck waiting for one of the consumers to release a state: 
```
				// Tracing state is used up, queue it for de-referencing. Note the
				// state is the parent state of trace block, use block.number-1 as
				// the state number.
				tracker.releaseState(task.block.NumberU64()-1, task.release)

				// Stream the result back to the result catcher or abort on teardown
				select {
				case resCh <- task:
				case <-closed:
					return
				}
```

Now, what I want to figure out, is wether it's at all possible for the following to happen: 

- The producer is stuck waiting for a state to be returned, 
- The consumers react to the `<-closed` event, and therefore do not drain the `taskCh`, and does not release the states sitting in `taskCh`, and thus do not un-stuck the producer.

**Comment by holiman** [comment]
128 max pending trace-states
8 threads

Producer feeds 8 states into the taskCh. 
One of the consumers  is really slow, but 7 of them run along

```
consumer-0: 1
consumer-1: 128
consumer-2: 127
consumer-3: 126
consumer-4: 125
consumer-5: 124
consumer-6: 123
consumer-7: 122
```

The producer will be stuck on `tracker.wait(129)`. 

At this point, if `closed` is triggered: All of the consumers will invoke `tracker.releaseState`, 
and thus freeing up the producer. The producer will then try to deliver the `129` state, 
but fail, and release that state in it's own loop. 

TLDR; All good! I don't see how a deadlock could occur here.

**Comment by holiman** [comment]
If you rebase again, the ubuntu appveyor build failure will be resolved.

**Comment by holiman** [comment]
The other failure is 
```
--- FAIL: TestEth2AssembleBlockWithAnotherBlocksTxs (0.04s)
    api_test.go:121: invalid number of transactions 0 != 1

```
If you want to get rid of that error, then :+1: on https://github.com/ethereum/go-ethereum/pull/25918 :)
